agent/src/com/cloud/agent/Agent.java

```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
             }
         }
     }
+
+    public class HostLBCheckerTask extends ManagedContextTimerTask {
+
+         @Override
+         protected void runInContext() {
+             try {
+                 int index = 0;
+                 final String[] msList = _shell.getHosts();
+                 String preferredHost  = msList[index];
+                 final String connectedHost = _shell.getConnectedHost();
+                 if (s_logger.isTraceEnabled()) {
+                     s_logger.trace("Running host LB checker task, connected host=" + connectedHost + ", preferred host=" + preferredHost);
+                 }
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
+                     if (_link != null) {
+                         boolean isHostUp = true;
+                         try (final Socket socket = new Socket()) {
+                             socket.connect(new InetSocketAddress(preferredHost, _shell.getPort()), 5000);
+                         } catch (final IOException e) {
+                             isHostUp = false;
+                             if (s_logger.isDebugEnabled()) {
+                                 s_logger.debug("Host: " + preferredHost + " is not reachable, trying to reconnect to next host...");
+                             }
+                         }
+                         if (isHostUp && _link != null) {
+                             long initialTime = System.currentTimeMillis();
+                             while (System.currentTimeMillis() - initialTime < wait && _inProgress.get() > 0) {
+                                 s_logger.debug("Waiting for active commands to be completed");
+                                 Thread.sleep(5000l);
+                             }
+                             if (_inProgress.get() == 0) {
+                                 if (s_logger.isDebugEnabled()) {
+                                     s_logger.debug("Preferred host is found to be reachable, trying to reconnect to: " + preferredHost);
+                                 }
+                                 _shell.resetHostCounter();
+                                 reconnect(_link);
+                             }
+                         }
+                     }
+                     index = (index + 1) % msList.length;
```

Comment:
@nvazquez don't change the preferred host to some other host, it's because we assume the first in the list is the `preferred` host. For example, consider this use-case when using the `static` algorithm the first item in the list could be IP of the LB (netscalar/haproxy etc) and the use of this mechanism is to ensure HA for the mgmt server when this LB is down. If we change the preferred host, it breaks our use-case.

Reply:
Thanks @rhtyd, I assumed background task should attempt reconnection to any available host starting from the preferred one at the beggining of the list and trying in order on 'host' list. I refactored this background task to attempt reconnection only to preferred host as you suggested, I think it simplifies the feature